test/jdk/com/sun/jdi/InterruptHangTest.java

```diff
@@ -28,93 +28,220 @@
+        Thread interruptorThread;
+        if (preciseMode) {
+            interruptorThread =
+                DebuggeeWrapper.newThread(new PreciseInterruptor(Thread.currentThread()));
             interruptorThread.start();
-            try {
-                sync.wait();
-            } catch (InterruptedException ee) {
-                System.out.println("Debuggee interruptee: interrupted before starting loop");
-            }
+        } else if (aggressiveMode) {
+            interruptorThread =
+                DebuggeeWrapper.newThread(new AggressiveInterruptor(Thread.currentThread()));
+            interruptorThread.start();
+        } else {
+            interruptorThread = null; // we are in "remote" interrupt mode
         }
 
         // Debugger will keep stepping thru this loop
-        for (int ii = 0; ii < 200; ii++) {
-            answer++;
+        for (int ii = 0; ii < INTERRUPTS_EXPECTED; ii++) {
+            boolean wasInterrupted = false;
             try {
-                // Give other thread a chance to run
+                // Give other thread a chance to interrupt
                 Thread.sleep(100);
             } catch (InterruptedException ee) {
-                System.out.println("Debuggee interruptee: interrupted at iteration: "
-                                   + ii);
+                answer++;
+                wasInterrupted = true;
+                boolean isInterrupted = Thread.currentThread().isInterrupted();
+                System.out.println("Debuggee interruptee(" + ii + "): isInterrupted(" + isInterrupted + ")");
+                if (preciseMode) {
+                    // When Thread.sleep() is interrupted, the interrupted status of the thread
+                    // should remain cleared (since InterruptedException was thrown), and the
+                    // the next interrupt won't come in until after the sync.notify() below.
+                    // Note this is not true for the aggressive and remote modes since
+                    // interrupts can come in at any time, even while we are handling
+                    // an intrrupt.
+                    if (isInterrupted) {
+                        throw new RuntimeException("Thread is interrupted but shouldn't be.");
+                    }
+                    synchronized(InterruptHangTarg.sync) {
+                        // Let the interruptor thread know it can start interrupting again
+                        sync.notify();
+                    }
+                }
+            }
+            // Every Thread.sleep() call should get interrupted
+            if (!wasInterrupted) {
+                throw new RuntimeException("Thread was never interrupted during sleep: " + ii);
+            }
+        }
+
+        if (interruptorThread != null) {
+            synchronized(InterruptHangTarg.sync) {
+                // Kill the interrupter thread
+                interruptorThread.interrupt();
             }
```

Comment:
Nit: Just a suggestion to consider replacing this `interruptorThread.interrupt()` with setting a status field, eg. `doStop`.  It can be done in the same synchronized block:
```
        synchronized (InterruptHangTarg.sync) {
                interruptorThread.doStop();
        }
```
Both `AggressiveInterruptor` and `PreciseInterruptor` can subclass the `InterruptorThread`, so you can  define the `interruptorThread` of type `InterruptorThread`. The class `InterruptorThread` can accumulate the common part of both interrupting threads.
To make synchronization work correctly the `Thread.sleep(5)` in the `AggressiveInterruptor` can be replaced with `InterruptHangTarg.sync.wait(5)` similarly as it is done in the `PreciseInterruptor` but as timed-wait.

The reason, I'm suggesting this is that using the `interrupt()` to interrupt `interruptedThread` adds a little confusion and not really necessary. This is in hope to make the code simpler.

Reply:
Using the interrupt is carried over from the original version of the test. I did consider replacing it, but opted to keep it in so there was one less change in the test to review. I can do as you suggested if you wish.

I did have them sharing a common superclass at one point, but the only thing they ended up having in common is the `interruptee` field. They still each needed their own constructor and the `run()` methods. Although the `run()` methods are very similar looking, it is hard to factor them into common code due to the placement of the `synchronize` and the `wait()` call.

I don't like the idea of a wait(5) that will always end out timing out. It's misleading. We aren't really waiting, we are sleeping, and would only have chosen wait() instead of sleep() to accommodate code sharing. It also would make the PreciseInterruptor have a wait() time of 5ms, which we don't want. In fact it can often take more than 5ms due to the single stepping that is going on at the same time. Keeping the code for each interruptor separate makes it clearer what each is attempting to do.

Reply:
I'm okay to keep it as is, especially as the `interrupt()` was used before your fixes.

I was thinking about something like below:
```
    abstract static class Interruptor implements Runnable {
        Thread interruptee;
        boolean doStop = false;
        abstract protected action();

        void doStop() { doStop = true; }

        public void run() {
            synchronized (InterruptHangTarg.sync) {
                while (!doStop) {
                    InterruptHangTarg.interruptsSent++;
                    interruptee.interrupt();
                    try {
                        action();
                    } catch (InterruptedException ee) {
                        System.out.println("Debuggee Interruptor: finished after " +
                                           InterruptHangTarg.interruptsSent + " iterrupts");
                        break;
                    }
                }
            }
        }
    }
    static class AggressiveInterruptor extends Interruptor {
        AggressiveInterruptor(Thread interruptee) {
            this.interruptee = interruptee;
        }
        protected action() {
            InterruptHangTarg.sync.wait(5);
        }
    }
    static class PreciseInterruptor extends Interruptor {
        AggressiveInterruptor(Thread interruptee) {
            this.interruptee = interruptee;
        }
        protected action() {
            InterruptHangTarg.sync.wait();
        }
    }
```

Reply:
The first version I did was just a single class that handled both interruptors, and looked similar to what you have above, except instead of a call to action() it instead had a conditional to decide what action to take based on the mode (precise or aggressvie).

The problem with this is the `sychronized` statement. Initially I thought it was benign for the aggressive case, and even had a comment saying it was unnecessary for the aggressive case, but doing so unconditionally made it easier to share code. However, I ran into some problem with it. I'm not exactly sure now what it was. I can see that the `synchronized` for interruptorThread.interrupt() call will end up deadlocking, although I don't recall that being the issue I had. For some reason I thought it was a problem at the start of the test, but I can't see what it might have been. Possibly due to other restructuring it no longer exists.

Reply:
> I can see that the synchronized for interruptorThread.interrupt() call will end up deadlocking...
It is one of the reasons I tried to suggest to get rid of the `interruptorThread.interrupt()`. :-)